mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp

```diff
@@ -318,6 +318,83 @@ Type Conv2dGenerator::getDataType(OpBuilder &builder) const {
   return dataType;
 }
 
+bool Conv2dGenerator::hasWorkspace(OpBuilder &builder) const {
+  // Decide if a workspace is needed.
+  // Preconditions:
+  // - data type: fp16
+  // - operation: backward weight conv2d.
+  // - use XDLOPS.
+  // - No need to pad along Gemm M/N/K dimension.
+  bool result = false;
+
+  if (config.operation.hasValue()) {
+    mlir::Type dataType = getDataType(builder);
+    if ((config.operation.getValue() == miopen::ConvOpType::BwdWeight) &&
+        config.xdlops && (dataType == builder.getF16Type())) {
+
+      // Logic to check if we need to pad along GemmM/N/K dimension for backward
+      // weight convolution.
+      // FIXME. Merge with logic used in the tuning process.
```

Comment:
Why do you need to execute a portion of the tuning logic beforehand? I think you only need to check if all preconditions are met. Then if it ended up to be a config that can't satisfy the below, you can let tuning step to reject it naturally.

Reply:
If any of the preconditions is not met:
- backward weight convolution
- fp16
- XDLOPS
- not padding along Gemm M/N/K dimension

we will fallback to the conv2d kernel that does not require a workspace.

The tuning parameters with or without workspace can be used interchangeably.

Reply:
One question: what breaks when there's gemmM/N/K padding?

Also, re duplicate tuning logic - I think what needs to happen there is to refactor that whole pile of code since we're poking around in there: make some sort of `ConvDefinition` struct that has all the relevant info and then both the generator and tuning code will operate on (or have a) `ConvDefinition` - we can also make `ConvDefinition` convertible to `ConvContext` for serialization purposes, but still.

That'd also allow us to factor out the gemm size computations that have been copied everywhere in `ConvToGemm`

Reply:
> The tuning parameters with or without workspace can be used interchangeably.

I know that when MIOpen is facing this sort of situation, they will expose both ops. Such that:
 - Both ops are candidate. 
    - One faster op is faster but requires more memory
    - Another slower op is slower but requires less memory
 - The compromise between speed and memory is something that client can choose in their specific scenarios. 
 - At the tuning time, they will choose tuning parameter independently for individual ops

I think the question goes back to whether you/we are okay with always choosing the absolutely most high performance kernel over any other constraint? If yes, then this chunk of precondition decision would be totally necessary and there's no other way around a decision like this in conv2d generator.

Reply:
> One question: what breaks when there's gemmM/N/K padding?

In `miopen-dialect`, if we deliberately allow padding cases using atomic add kernel, the result would be 0 everywhere on both fp32/fp16 cases. The root cause is not known yet.


> Also, re duplicate tuning logic - I think what needs to happen there is to refactor that whole pile of code since we're poking around in there: make some sort of `ConvDefinition` struct that has all the relevant info and then both the generator and tuning code will operate on (or have a) `ConvDefinition` - we can also make `ConvDefinition` convertible to `ConvContext` for serialization purposes, but still.
> 
> That'd also allow us to factor out the gemm size computations that have been copied everywhere in `ConvToGemm`

Agree an overhaul here is necessary, I see somewhat duplicated structures like `Conv2dGenerator::Config` in the generator directory, and `ConvContext` in the tuning domain.

Reply:
And the advantage to the overhaul is that we could memoize the gemm size computation and tuning selection, instead of running that valid tuning parameter section loop a bunch of times.

So yeah, I'd rather this not go in without an overhaul to get rid of this copy-paste.

And can we re-test the padding thing, just to see if something changed in the meantime?

Reply:
The objection I have is that the temporary copy-paste may become much less temporary because new features will get prioritized over fixing this

Reply:
That's why we use tickets to track the laundry list of known todo items.

Reply:
https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/439 to keep track of the new requirements identified.

Reply:
So the dilemma we face here is:
1) `conv2dGenerator` is not a tuner (ex: solver within MIOpen), it possess zero knowledge of the tuning search space.
2) `conv2dGenerator` does not know the internals of the tuning process (M/N/KPerBlock, M/NPerWave, KPACK, etc).

But, to determine whether padding kernel is used requires the knowledge of the tuning parameters possibly be used for a particular convolution configuration.

As a result, in the current form of the PR, since `conv2dGenerator` doesn't know the tuning search space (1 above), it has to be amended to possess part of 2. Therefore we have duplicated logic.

Reply:
@whchung Actually inspired by what you are doing here. Wouldn't it be possible to derive the tuning parameter in constructing the conv op:
 - In constructing the conv op, derive tuning parameters before a conv op is constructed
   - If succeed, the constructed conv op would have tuning parameters, so there's no need to do it as a lowering step
   - If failed, we'd return a empty handle, allowing the failure to happen even earlier
 - It sounds like if all tuning parameters validation are brought earlier, we can (use the info to) decide what kind of conv op to construct

Reply:
To remedy duplicated logic, I'd like create an oracle function to judge whether a padding kernel is needed, given the GemmM/N/K size, and the direction of the convolution.

The oracle function would possess the vector of initial tuning parameters, just like what's in use in `GridwiseGemmParams`.

On the other hand, I will NOT consolidate somewhat duplicated structures between generator and tuning modules: `Conv2dGenerator::Config` / `ConvContext`.

Reply:
@jerryyin You are talking about eliminating `-miopen-affix-params` as an independent pass and run it inside `conv2dGenerator`? I'm not against the idea but it still seems to be beyond the scope of the PR. I'll log it in https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/439

Reply:
I agree with @jerryyin's proposal, more or less (One interesting thing to consider is whether a convOp can hold the set of valid configurations - or if it could generate it - so that  tuning loops can be told what the valid parameters are for a particular configuration)

Reply:
Initial version of refactoring is done. `calculatePaddingKernelSize` is made as a function instead a lambda.